### PR TITLE
Openshift3 deployment fixes

### PR DESCRIPTION
**Pull request by alesrosina**
There where some leftovers from @GrahamDumpleton 's fork, so I started with some small fixes, since in [last openshift issue](https://github.com/WeblateOrg/weblate/issues/1523) he mentioned he is finished with this. 

Otherwise this MR is a start of issue https://github.com/WeblateOrg/weblate/issues/2476

### CODE REVIEW

**Comment by GrahamDumpleton** [comment]
FWIW. There are better ways than relying on ``nss_wrapper`` now.

**Comment by alesrosina** [comment]
@GrahamDumpleton do you have any better suggestion? I didn't find anything else and this is my first time playing with openshift so would appreciate your input.

**Comment by GrahamDumpleton** [comment]
Until I can show a tailored example, have a dig through what Jupyter notebook image does.

* https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile#L47
* https://github.com/jupyter/docker-stacks/blob/master/base-notebook/start.sh#L112

Their case is slightly different as they don't make things group ``root``. Having group ``root`` is better option.

Steps which are different from using ``nss_wrapper`` are:

* Make ``/etc/passwd`` group writable.
* In startup, add user entry to ``/etc/passwd`` for user ID if it doesn't exist.
* Add ``wheel`` group if doesn't exist.
* Update PAM so that only users in group ``wheel`` can use ``su``.

The ``su`` stuff is needed to prevent someone adding a password for ``root`` to the ``/etc/passwd`` file and then using ``su``. That restricts it to a user who is in group ``wheel`` which wouldn't normally be the case. This is more for use outside of OpenShift, as OpenShift doesn't allow ``su`` anyway. But then I just realised another potential problem with this if image is used outside of OpenShift, which I will need to look into.

So right now ``nss_wrapper`` is still the safest.

**Comment by alesrosina** [comment]
@GrahamDumpleton ok, so I tried cupple of things, but I am failing badly :) Do you have any idea hot to install it? Do I really need to create my own docker image, that already includes this? Otherwise I'm always bumping into permission error, since `yum install nss_wrapper`needs to run under sudo ...

I would like to avoid creating my own docker image, since that means one maintenance more for weblate.

**Comment by GrahamDumpleton** [comment]
Now I am confused. I was assuming that since you were playing around with ``nss_wrapper`` that you were trying to fix some existing ``Dockerfile`` type build. If you are using an OpenShift Python S2I build, you don't need to fiddle with ``nss_wrapper`` since the base image already includes it and does what is required for the application. So right now I have no idea what problem you are trying to solve. Can you point to the specific issue which explains in detail what you current problem is. The only issue I can find doesn't include a sufficient explanation of how/where ``git clone`` is called.

**Comment by alesrosina** [comment]
Yes, so I used your implementation for openshift, so I ran this command:
`oc new-app https://raw.githubusercontent.com/WeblateOrg/weblate/master/openshift3/persistent-postgresql.json`

Application is running, but there is a functionality for generating SSH key in weblate, which returns `Failed to generate key: b'No user exists for uid 1000140000\r\n'`error. As far as I understood, this is a "feature" of openshift, so you need to use `nss_wrapper` for that.

But as far as I can see, you are using standard `python:3.5` docker image or am I missing sth? Where is this `nss_wrapper` already included?

**Comment by GrahamDumpleton** [comment]
Where in the weblate code is the SSH key gen command called? That is what I need to know. Was there a Python stack trace showing where in the code it was being done from?

**Comment by alesrosina** [comment]
I am guessing it is happening here: https://github.com/WeblateOrg/weblate/blob/e78de11db7b2beefb169d876a2b2034ceec52e21/weblate/wladmin/views.py#L78

But maybe @nijel can explain a bit more into detail what is going on and how this ssh thing works.

**Comment by alesrosina** [comment]
ok, actually is this function that gets called: 
https://github.com/WeblateOrg/weblate/blob/e78de11db7b2beefb169d876a2b2034ceec52e21/weblate/vcs/ssh.py#L190

So it will execute `ssh-keygen` on openshift. And yeah, if I run this in console on openshit, got exactly that error.

**Comment by GrahamDumpleton** [comment]
The problem is this line:

* https://github.com/WeblateOrg/weblate/blob/39ae3e37edeb6890fae619a559164602da7d7f58/weblate/vcs/ssh.py#L123

Because it clears environment, it isn't passing through the ``LD_PRELOAD`` and ``NSS_*`` environment variables. This means that the ``ssh-keygen`` program will not inherit the fiddle required to use the alternate ``passwd`` file.

**Comment by GrahamDumpleton** [comment]
The question is whether a change to:

* https://github.com/WeblateOrg/weblate/blob/39ae3e37edeb6890fae619a559164602da7d7f58/weblate/trans/util.py#L166

is going to be acceptable, to also pass ``LD_PRELOAD`` and the ``NSS_*`` environment variables (can't remember names of these), given that is an OpenShift specific issue.

**Comment by GrahamDumpleton** [comment]
These two variables:

* https://github.com/sclorg/s2i-python-container/blob/master/2.7/root/opt/app-root/etc/generate_container_user#L14

and ``LD_PRELOAD``.

**Comment by alesrosina** [comment]
@GrahamDumpleton yes, this works now. Thank you very much for this help, I would struggle much longer if you would not point me to this direction.

@nijel i don't see a connection with failing AppVeyor, maybe you can check this? And also check if you agree with change.

For adding cashing, I will add a new comit, so let's keep this MR in WIP state.

**Comment by nijel** [comment]
You can ignore failing AppVeyor, we're not yet there on passing tests there, see https://github.com/WeblateOrg/weblate/issues/2267.

File: weblate/trans/util.py [hunk 1]
```diff
@@ -171,7 +171,14 @@ def get_clean_env(extra=None):
...
         environ.update(extra)
-    variables = ('PATH', 'LD_LIBRARY_PATH', 'SystemRoot')
+    variables = (
+        'PATH',
+        'LD_LIBRARY_PATH',
+        'NSS_WRAPPER_GROUP',
+        'NSS_WRAPPER_PASSWD',
+        'LD_PRELOAD',
+        'SystemRoot',
+      )
```
Comment (nijel) [inline weblate/trans/util.py]: The list probably grew up too much, so it might be good to document why each variable is needed.

Comment (alesrosina) [reply weblate/trans/util.py]: OK, added a comment. Do you find it readable enough?

**Comment by nijel** [comment]
Merged, thanks for your contribution!

**Pull request merged**